### PR TITLE
Introduce BBJ_CALLFINALLYRET block kind

### PR DESCRIPTION
**Pull request by BruceForstall**
Non-exceptional calls to finally blocks are represented in the JIT IR
as a pair of blocks:
1. `BBJ_CALLFINALLY` targets the finally block
2. `BBJ_ALWAYS` targets the finally "continuation", namely, the location
where execution should continue after calling the finally. It is different
from normal `BBJ_ALWAYS`, as it must be empty of code, must be the `bbNext`
successor of the `BBJ_CALLFINALLY`, and must have the
`BBF_KEEP_BBJ_ALWAYS` flag.

If the JIT knows the finally never returns (e.g., it unconditionally
throws), the paired `BBJ_ALWAYS` will not exist, and the `BBJ_CALLFINALLY`
will have the `BBF_RETLESS_CALL` flag.

This structure is unique in the JIT IR and leads to a lot of special
case code to maintain the pair relationship.

This PR makes a small change to the IR by introducing a new `BBJ_CALLFINALLYRET`
block type to be used instead of `BBJ_ALWAYS` when paired with a `BBJ_CALLFINALLY`.
The idea is to reduce overloading the semantics of `BBJ_ALWAYS` in this case,
so code that checks for `BBJ_ALWAYS` can do so knowing it always means the same
thing, without checking whether it is one of the special pair.

There one remaining "special" `BBJ_ALWAYS`: the "final step" block of an x86
callfinally. It is the only block that still uses the `BBF_KEEP_BBJ_ALWAYS` flag.

This change has a very few diffs where a slight difference in the order of
block removal (especially, in `fgRemoveEmptyTry`) leads to different
downstream optimizations.

Currently, this is implemented by running a pass to convert paired `BBJ_ALWAYS`
to `BBJ_CALLFINALLYRET` after the importer (namely, after importing `LEAVE`).
I'll go back and fix the LEAVE implementation if this is merged, to generate
the new IR form directly in the importer, and remove a few `TODO-Quirk` left
as a result (e.g., in `fgFindInsertPoint`, `BBPredsChecker::CheckEhTryDsc`,
`fgComputeMissingBlockWeights`).

The benefit of this change is the IR is more explicit; asserts can be better;
and there should be less confusion about `BBJ_ALWAYS` cases "accidentally"
kicking in for callfinally pairs.

Addresses #95355

### CODE REVIEW

**Comment by BruceForstall** [comment]
@AndyAyersMS PTAL
cc @dotnet/jit-contrib

**Review by AndyAyersMS — APPROVED** [review]
Just saw a few small things, feel free to defer addressing these.

File: src/coreclr/jit/fgehopt.cpp [hunk 1]
```diff
@@ -147,40 +147,33 @@ PhaseStatus Compiler::fgRemoveEmptyFinally()
...
                         postTryFinallyBlock->bbNum, leaveBlock->bbNum);
 
-                noway_assert(leaveBlock->KindIs(BBJ_ALWAYS));
+                // Remove the `leaveBlock` first.
+                nextBlock = leaveBlock->Next();
+                fgPrepareCallFinallyRetForRemoval(leaveBlock);
+                fgRemoveBlock(leaveBlock, /* unreachable */ true);
 
                 currentBlock->SetKindAndTarget(BBJ_ALWAYS, postTryFinallyBlock);
+                currentBlock->RemoveFlags(BBF_RETLESS_CALL); // no longer a BBJ_CALLFINALLY
```
[the remark this answers is not in the document]
Comment (BruceForstall) [reply src/coreclr/jit/fgehopt.cpp:171]: What happens is the code above to remove the `leaveBlock` adds the retless flag, so here we just remove it.

**Comment by BruceForstall** [comment]
[Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=500403&view=ms.vss-build-web.run-extensions-tab). Very small asmdiffs, but there is a TP impact up to +0.10% -- not surprising, given some extra conditionals.

**Pull request merged**